## Canada Neighborhood Analysis project

This notebook aims to get neighborhoods data from wikipedia and then leverage the foursquare API to perform segmentation and clustering analysis

Link to wikipedia article: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M



## Part 1: Getting the dataset from wikipedia using pandas

In [1]:
import pandas as pd


In [2]:
#Resetting default display limit on pandas dataframes so we don't have to deal with ellipses splitting the dataset in the outputs
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', -1)

#Using the read_html library from python to pull the raw table
df=pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]
df.head()

Postal code           Borough                Neighborhood
0  M1A         Not assigned      NaN                       
1  M2A         Not assigned      NaN                       
2  M3A         North York        Parkwoods                 
3  M4A         North York        Victoria Village          
4  M5A         Downtown Toronto  Regent Park / Harbourfront

In [3]:
#Exploring entire dataset
df

Postal code           Borough  \
0    M1A         Not assigned       
1    M2A         Not assigned       
2    M3A         North York         
3    M4A         North York         
4    M5A         Downtown Toronto   
5    M6A         North York         
6    M7A         Downtown Toronto   
7    M8A         Not assigned       
8    M9A         Etobicoke          
9    M1B         Scarborough        
10   M2B         Not assigned       
11   M3B         North York         
12   M4B         East York          
13   M5B         Downtown Toronto   
14   M6B         North York         
15   M7B         Not assigned       
16   M8B         Not assigned       
17   M9B         Etobicoke          
18   M1C         Scarborough        
19   M2C         Not assigned       
20   M3C         North York         
21   M4C         East York          
22   M5C         Downtown Toronto   
23   M6C         York               
24   M7C         Not assigned       
25   M8C         Not assigned       
26   M9C         Etobicoke          
27   M1E         Scarborough        
28   M2E         Not assigned       
29   M3E         Not assigned       
30   M4E         East Toronto       
31   M5E         Downtown Toronto   
32   M6E         York               
33   M7E         Not assigned       
34   M8E         Not assigned       
35   M9E         Not assigned       
36   M1G         Scarborough        
37   M2G         Not assigned       
38   M3G         Not assigned       
39   M4G         East York          
40   M5G         Downtown Toronto   
41   M6G         Downtown Toronto   
42   M7G         Not assigned       
43   M8G         Not assigned       
44   M9G         Not assigned       
45   M1H         Scarborough        
46   M2H         North York         
47   M3H         North York         
48   M4H         East York          
49   M5H         Downtown Toronto   
50   M6H         West Toronto       
51   M7H         Not assigned       
52   M8H         Not assigned       
53   M9H         Not assigned       
54   M1J         Scarborough        
55   M2J         North York         
56   M3J         North York         
57   M4J         East York          
58   M5J         Downtown Toronto   
59   M6J         West Toronto       
60   M7J         Not assigned       
61   M8J         Not assigned       
62   M9J         Not assigned       
63   M1K         Scarborough        
64   M2K         North York         
65   M3K         North York         
66   M4K         East Toronto       
67   M5K         Downtown Toronto   
68   M6K         West Toronto       
69   M7K         Not assigned       
70   M8K         Not assigned       
71   M9K         Not assigned       
72   M1L         Scarborough        
73   M2L         North York         
74   M3L         North York         
75   M4L         East Toronto       
76   M5L         Downtown Toronto   
77   M6L         North York         
78   M7L         Not assigned       
79   M8L         Not assigned       
80   M9L         North York         
81   M1M         Scarborough        
82   M2M         North York         
83   M3M         North York         
84   M4M         East Toronto       
85   M5M         North York         
86   M6M         York               
87   M7M         Not assigned       
88   M8M         Not assigned       
89   M9M         North York         
90   M1N         Scarborough        
91   M2N         North York         
92   M3N         North York         
93   M4N         Central Toronto    
94   M5N         Central Toronto    
95   M6N         York               
96   M7N         Not assigned       
97   M8N         Not assigned       
98   M9N         York               
99   M1P         Scarborough        
100  M2P         North York         
101  M3P         Not assigned       
102  M4P         Central Toronto    
103  M5P         Central Toronto    
104  M6P         West Toronto       
105  M7P         Not assigned       
106  M8P         Not assigned       
107  M9P

In [4]:
#Dropping rows where Borough is Not Assigned
indexnames= df[df['Borough']=="Not assigned"].index
df.drop(indexnames, inplace=True)

#Replace "/" with "," in the neighborhoods column
df['Neighborhood']=df['Neighborhood'].str.replace('/',',')

#Display dataframe
df

# We see from the output below that there are no boroughs that do not have at least one neighborhood assigned to it. 
#So we don't need to rund the code to assign borough name to neighborhood name where neighborhood field is blank





Postal code           Borough  \
2    M3A         North York         
3    M4A         North York         
4    M5A         Downtown Toronto   
5    M6A         North York         
6    M7A         Downtown Toronto   
8    M9A         Etobicoke          
9    M1B         Scarborough        
11   M3B         North York         
12   M4B         East York          
13   M5B         Downtown Toronto   
14   M6B         North York         
17   M9B         Etobicoke          
18   M1C         Scarborough        
20   M3C         North York         
21   M4C         East York          
22   M5C         Downtown Toronto   
23   M6C         York               
26   M9C         Etobicoke          
27   M1E         Scarborough        
30   M4E         East Toronto       
31   M5E         Downtown Toronto   
32   M6E         York               
36   M1G         Scarborough        
39   M4G         East York          
40   M5G         Downtown Toronto   
41   M6G         Downtown Toronto   
45   M1H         Scarborough        
46   M2H         North York         
47   M3H         North York         
48   M4H         East York          
49   M5H         Downtown Toronto   
50   M6H         West Toronto       
54   M1J         Scarborough        
55   M2J         North York         
56   M3J         North York         
57   M4J         East York          
58   M5J         Downtown Toronto   
59   M6J         West Toronto       
63   M1K         Scarborough        
64   M2K         North York         
65   M3K         North York         
66   M4K         East Toronto       
67   M5K         Downtown Toronto   
68   M6K         West Toronto       
72   M1L         Scarborough        
73   M2L         North York         
74   M3L         North York         
75   M4L         East Toronto       
76   M5L         Downtown Toronto   
77   M6L         North York         
80   M9L         North York         
81   M1M         Scarborough        
82   M2M         North York         
83   M3M         North York         
84   M4M         East Toronto       
85   M5M         North York         
86   M6M         York               
89   M9M         North York         
90   M1N         Scarborough        
91   M2N         North York         
92   M3N         North York         
93   M4N         Central Toronto    
94   M5N         Central Toronto    
95   M6N         York               
98   M9N         York               
99   M1P         Scarborough        
100  M2P         North York         
102  M4P         Central Toronto    
103  M5P         Central Toronto    
104  M6P         West Toronto       
107  M9P         Etobicoke          
108  M1R         Scarborough        
109  M2R         North York         
111  M4R         Central Toronto    
112  M5R         Central Toronto    
113  M6R         West Toronto       
114  M7R         Mississauga        
116  M9R         Etobicoke          
117  M1S         Scarborough        
120  M4S         Central Toronto    
121  M5S         Downtown Toronto   
122  M6S         West Toronto       
126  M1T         Scarborough        
129  M4T         Central Toronto    
130  M5T         Downtown Toronto   
135  M1V         Scarborough        
138  M4V         Central Toronto    
139  M5V         Downtown Toronto   
142  M8V         Etobicoke          
143  M9V         Etobicoke          
144  M1W         Scarborough        
147  M4W         Downtown Toronto   
148  M5W         Downtown Toronto   
151  M8W         Etobicoke          
152  M9W         Etobicoke          
153  M1X         Scarborough        
156  M4X         Downtown Toronto   
157  M5X         Downtown Toronto   
160  M8X         Etobicoke          
165  M4Y         Downtown Toronto   
168  M7Y         East Toronto       
169  M8Y         Etobicoke          
178  M8Z         Etobicoke          

                                                                                                                                      Neighborhood  
2    P

In [5]:
#Resetting Index
df.reset_index(inplace = True, drop = True)
df.head()

Postal code           Borough                                  Neighborhood
0  M3A         North York        Parkwoods                                   
1  M4A         North York        Victoria Village                            
2  M5A         Downtown Toronto  Regent Park , Harbourfront                  
3  M6A         North York        Lawrence Manor , Lawrence Heights           
4  M7A         Downtown Toronto  Queen's Park , Ontario Provincial Government

In [6]:
df.shape

(103, 3)

It seems like we have 103 rows in our data set

## Part 2: Getting Latitude and Longitude codes for the neighborhoods

In [7]:
#installing geocoder

#!conda install -c conda-forge geocoder

In [8]:
# using CSV file to get latitude and longitude data
df_coord= pd.read_csv('http://cocl.us/Geospatial_data')

#Renaming Postal Code to Postal code in this dataset
df_coord.rename(columns = {'Postal Code':'Postal code'}, inplace = True) 

#Filtering only rows with torontoin the bourough name
df_canada=df[df['Borough'].str.contains("Toronto")]

# joining using postal codes to dataframe with name df_canada ( To be later filtered with only toronto neighborhoods)
df_toronto=pd.merge(left=df_canada, right=df_coord, left_on='Postal code', right_on='Postal code')

df_toronto.head()

Postal code           Borough                                  Neighborhood  \
0  M5A         Downtown Toronto  Regent Park , Harbourfront                     
1  M7A         Downtown Toronto  Queen's Park , Ontario Provincial Government   
2  M5B         Downtown Toronto  Garden District, Ryerson                       
3  M5C         Downtown Toronto  St. James Town                                 
4  M4E         East Toronto      The Beaches                                    

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

# Part3: Exploring and Clustering Toronto neighborhoods

In [9]:
# Importing additional libraries for our neigborhood analysis
import json # library to handle JSON files


from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes 

print('Libraries imported.')

Libraries imported.


In [10]:
#!conda install -c conda-forge folium=0.5.0 --yes 
# !pip install folium

import folium

In [11]:
# Getting coordinates of Toronto to build the map
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [12]:
!pip install folium

In [13]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

 Defining a Function to get Nearby Venues for each neighborhood in the dataframe

In [14]:
#Foursquare credentials:

CLIENT_ID = 'CDYEYZFDWHQRZP4R2PQE1VTVWF512VW4XVTIQCYICPOLBRNK' # your Foursquare ID
CLIENT_SECRET = 'IQV5PHHFTNCQ4CM4YTV0D15HVAQAR3TEUQ1GLQA0OM4HTOWV' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

In [15]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
       
    
        # create the API request URL
        
        LIMIT = 100

        radius= 500
        
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [16]:
#Getting nearby venues for each neighborhood in the dataframe
toronto_venues = getNearbyVenues(names=df_toronto['Neighborhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

Regent Park , Harbourfront
Queen's Park , Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond , Adelaide , King
Dufferin , Dovercourt Village
Harbourfront East , Union Station , Toronto Islands
Little Portugal , Trinity
The Danforth West , Riverdale
Toronto Dominion Centre , Design Exchange
Brockton , Parkdale Village , Exhibition Place
India Bazaar , The Beaches West
Commerce Court , Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park , The Junction South
North Toronto West
The Annex , North Midtown , Yorkville
Parkdale , Roncesvalles
Davisville
University of Toronto , Harbord
Runnymede , Swansea
Moore Park , Summerhill East
Kensington Market , Chinatown , Grange Park
Summerhill West , Rathnelly , South Hill , Forest Hill SE , Deer Park
CN Tower , King and Spadina , Railway Lands , Harbourfront West , Bathurst Quay , South Niagara , Island airport
Roseda

In [17]:
# Checking dimensions of dataframe
print(toronto_venues.shape)
toronto_venues.head()

(1602, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park , Harbourfront  43.65426              -79.360636                
1  Regent Park , Harbourfront  43.65426              -79.360636                
2  Regent Park , Harbourfront  43.65426              -79.360636                
3  Regent Park , Harbourfront  43.65426              -79.360636                
4  Regent Park , Harbourfront  43.65426              -79.360636                

                    Venue  Venue Latitude  Venue Longitude  \
0  Roselle Desserts        43.653447      -79.362017         
1  Tandem Coffee           43.653559      -79.361809         
2  Cooper Koo Family YMCA  43.653249      -79.358008         
3  Body Blitz Spa East     43.654735      -79.359874         
4  Morning Glory Cafe      43.653947      -79.361149         

        Venue Category  
0  Bakery               
1  Coffee Shop          
2  Distribution Center  
3  Spa                  
4  Breakfast Spot

In [18]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                                                                              
Berczy Park                                                                                                       56                      
Brockton , Parkdale Village , Exhibition Place                                                                    23                      
Business reply mail Processing CentrE                                                                             14                      
CN Tower , King and Spadina , Railway Lands , Harbourfront West , Bathurst Quay , South Niagara , Island airport  18                      
Central Bay Street                                                                                                59                      
Christie                                                                                                          18                      
Church and Wellesley                                                                                              73                      
Commerce Court , Victoria Hotel                                                                                   100                     
Davisville                                                                                                        34                      
Davisville North                                                                                                  7                       
Dufferin , Dovercourt Village                                                                                     18                      
First Canadian Place , Underground city                                                                           100                     
Forest Hill North & West                                                                                          4                       
Garden District, Ryerson                                                                                          100                     
Harbourfront East , Union Station , Toronto Islands                                                               100                     
High Park , The Junction South                                                                                    23                      
India Bazaar , The Beaches West                                                                                   20                      
Kensington Market , Chinatown , Grange Park                                                                       57                      
Lawrence Park                                                                                                     3                       
Little Portugal , Trinity                                                                                         42                      
Moore Park , Summerhill East                                                                                      2                       
North Toronto West                                                                                                18                      
Parkdale , Roncesvalles                                                                                           13                      
Queen's Park , Ontario Provincial Government                                                                      35                      
Regent Park , Harbourfront                                                                                        48                      
Richmond , Adelaide , King                                                                                        94                      
Rosedale                                                                                                          4                       
Roselawn                                                                           

Determining unique categories that can be created from data returned by foursquare

In [19]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 228 uniques categories.


**Analyzing the Neighborhoods**

In [20]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0  0            0        0                   0             0                
1  0            0        0                   0             0                
2  0            0        0                   0             0                
3  0            0        0                   0             0                
4  0            0        0                   0             0                

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0  0                0                 0                    0              
1  0                0                 0                    0              
2  0                0                 0                    0              
3  0                0                 0                    0              
4  0                0                 0                    0              

   Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0  0         0            0                    0                  
1  0         0            0                    0                  
2  0         0            0                    0                  
3  0         0            0                    0                  
4  0         0            0                    0                  

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0  0                   0              0          0           0            
1  0                   0              0          0           0            
2  0                   0              0          0           0            
3  0                   0              0          0           0            
4  0                   0              0          0           0            

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0  1       0     0    0                 0                   0       
1  0       0     0    0                 0                   0       
2  0       0     0    0                 0                   0       
3  0       0     0    0                 0                   0       
4  0       0     0    0                 0                   0       

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0  0                0         0           0                   0        
1  0                0         0           0                   0        
2  0                0         0           0                   0        
3  0                0         0           0                   0        
4  0                0         0           0                   0        

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0  0              0          0         0                     0                
1  0              0          0         0                     0                
2  0              0          0         0                     0                
3  0              0          0         0                     0                
4  0              0          0         0                     1                

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0  0        0                0         0             0              0          
1  0        0                0         0             0              0          
2  0        0                0         0             0              0          
3  0        0                0         0             0              0          
4  0        0                0         0             0              0          

   Butcher  Cable Car  Café  Cajun / Creole Restaurant  Candy Store  \
0  0        0          0     0                          0             
1  0        0          0     0                          0             
2  0        0          0     0                          0             
3  0        0          0     0                          0             
4  0        0          0     0                          0             

   C

In [21]:
# Rearranging columns using a loop instead

toronto_onehot= toronto_onehot[['Neighborhood'] + [ col for col in toronto_onehot.columns if col != 'Neighborhood' ] ]
toronto_onehot.head()

Neighborhood  Yoga Studio  Airport  Airport Food Court  \
0  Regent Park , Harbourfront  0            0        0                    
1  Regent Park , Harbourfront  0            0        0                    
2  Regent Park , Harbourfront  0            0        0                    
3  Regent Park , Harbourfront  0            0        0                    
4  Regent Park , Harbourfront  0            0        0                    

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0  0             0               0                0                  
1  0             0               0                0                  
2  0             0               0                0                  
3  0             0               0                0                  
4  0             0               0                0                  

   American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0  0                    0             0         0             
1  0                    0             0         0             
2  0                    0             0         0             
3  0                    0             0         0             
4  0                    0             0         0             

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0  0                    0                 0                   0               
1  0                    0                 0                   0               
2  0                    0                 0                   0               
3  0                    0                 0                   0               
4  0                    0                 0                   0               

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0  0          0           0           1       0     0    0                  
1  0          0           0           0       0     0    0                  
2  0          0           0           0       0     0    0                  
3  0          0           0           0       0     0    0                  
4  0          0           0           0       0     0    0                  

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0  0                   0      0                0         0            
1  0                   0      0                0         0            
2  0                   0      0                0         0            
3  0                   0      0                0         0            
4  0                   0      0                0         0            

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0  0                   0       0              0          0          
1  0                   0       0              0          0          
2  0                   0       0              0          0          
3  0                   0       0              0          0          
4  0                   0       0              0          0          

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0  0                     0               0        0                0          
1  0                     0               0        0                0          
2  0                     0               0        0                0          
3  0                     0               0        0                0          
4  0                     1               0        0                0          

   Burger Joint  Burrito Place  Bus Line  Butcher  Cable Car  Café  \
0  0             0              0         0        0          0      
1  0             0              0         0        0          0      
2  0             0              0         0        0          0      
3  0             0              0         0        0          0      
4  0             0              0         0        0          0      

   Cajun / Creole Restaurant  Candy Store  Caribbean Restaurant  Cheese Shop  \
0  0      

In [22]:
# Examining Data Frame Size
print(toronto_onehot.shape)

#Grouping rows by neigborhood and calculating means of frequencies of venue categories
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

(1602, 228)


Neighborhood  \
0   Berczy Park                                                                                                        
1   Brockton , Parkdale Village , Exhibition Place                                                                     
2   Business reply mail Processing CentrE                                                                              
3   CN Tower , King and Spadina , Railway Lands , Harbourfront West , Bathurst Quay , South Niagara , Island airport   
4   Central Bay Street                                                                                                 
5   Christie                                                                                                           
6   Church and Wellesley                                                                                               
7   Commerce Court , Victoria Hotel                                                                                    
8   Davisville                                                                                                         
9   Davisville North                                                                                                   
10  Dufferin , Dovercourt Village                                                                                      
11  First Canadian Place , Underground city                                                                            
12  Forest Hill North & West                                                                                           
13  Garden District, Ryerson                                                                                           
14  Harbourfront East , Union Station , Toronto Islands                                                                
15  High Park , The Junction South                                                                                     
16  India Bazaar , The Beaches West                                                                                    
17  Kensington Market , Chinatown , Grange Park                                                                        
18  Lawrence Park                                                                                                      
19  Little Portugal , Trinity                                                                                          
20  Moore Park , Summerhill East                                                                                       
21  North Toronto West                                                                                                 
22  Parkdale , Roncesvalles                                                                                            
23  Queen's Park , Ontario Provincial Government                                                                       
24  Regent Park , Harbourfront                                                                                         
25  Richmond , Adelaide , King                                                                                         
26  Rosedale                                                                                                           
27  Roselawn                                                                                                           
28  Runnymede , Swansea                                                                                                
29  St. James Town                                                                                                     
30  St. James Town , Cabbagetown                                                                                       
31  Stn A PO Boxes                                                                                                     
32  Studio District                                                                                                    
33  Summerhill West , Ra

In [23]:
#Creating DataFrame to check top 10 venues in each neighborhood

#defining function to return most common venues
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

import numpy as np

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0  Berczy Park                                                                                                        
1  Brockton , Parkdale Village , Exhibition Place                                                                     
2  Business reply mail Processing CentrE                                                                              
3  CN Tower , King and Spadina , Railway Lands , Harbourfront West , Bathurst Quay , South Niagara , Island airport   
4  Central Bay Street                                                                                                 

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Cocktail Bar          Coffee Shop           Farmers Market         
1  Café                  Breakfast Spot        Coffee Shop            
2  Pizza Place           Auto Workshop         Skate Park             
3  Airport Service       Airport Lounge        Airport Terminal       
4  Coffee Shop           Italian Restaurant    Sandwich Place         

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0  Seafood Restaurant    Bakery                Restaurant             
1  Pet Store             Stadium               Bar                    
2  Comic Shop            Brewery               Gym / Fitness Center   
3  Coffee Shop           Boutique              Bar                    
4  Café                  Bubble Tea Shop       Salad Place            

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0  Italian Restaurant    Café                  Cheese Shop            
1  Restaurant            Bakery                Intersection           
2  Light Rail Station    Garden Center         Garden                 
3  Rental Car Location   Sculpture Garden      Plane                  
4  Japanese Restaurant   Burger Joint          Ice Cream Shop         

  10th Most Common Venue  
0  Beer Bar               
1  Italian Restaurant     
2  Restaurant             
3  Boat or Ferry          
4  Yoga Studio

**K-means Clustering

In [24]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [25]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal code           Borough                                  Neighborhood  \
0  M5A         Downtown Toronto  Regent Park , Harbourfront                     
1  M7A         Downtown Toronto  Queen's Park , Ontario Provincial Government   
2  M5B         Downtown Toronto  Garden District, Ryerson                       
3  M5C         Downtown Toronto  St. James Town                                 
4  M4E         East Toronto      The Beaches                                    

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636  0               Coffee Shop            
1  43.662301 -79.389494  0               Coffee Shop            
2  43.657162 -79.378937  0               Coffee Shop            
3  43.651494 -79.375418  0               Café                   
4  43.676357 -79.293031  0               Coffee Shop            

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Pub                   Bakery                Park                   
1  Sushi Restaurant      Diner                 College Auditorium     
2  Clothing Store        Café                  Cosmetics Shop         
3  Coffee Shop           Cocktail Bar          Gastropub              
4  Asian Restaurant      Trail                 Pub                    

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0  Restaurant            Breakfast Spot             Theater                
1  Spa                   Beer Bar                   Sandwich Place         
2  Japanese Restaurant   Middle Eastern Restaurant  Bubble Tea Shop        
3  Italian Restaurant    American Restaurant        Hotel                  
4  Health Food Store     Women's Store              Dog Run                

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Café                  Beer Store            Ice Cream Shop         
1  Burger Joint          Burrito Place         Café                   
2  Tea Room              Fast Food Restaurant  Ramen Restaurant       
3  Seafood Restaurant    Moroccan Restaurant   Cosmetics Shop         
4  Diner                 Discount Store        Distribution Center

In [26]:
import folium

**DISPLAYING THE CLUSTER MAP**

In [27]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters

Cluster 1: 

In [31]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  \
0   Regent Park , Harbourfront                                                                                         
1   Queen's Park , Ontario Provincial Government                                                                       
2   Garden District, Ryerson                                                                                           
3   St. James Town                                                                                                     
4   The Beaches                                                                                                        
5   Berczy Park                                                                                                        
6   Central Bay Street                                                                                                 
7   Christie                                                                                                           
8   Richmond , Adelaide , King                                                                                         
9   Dufferin , Dovercourt Village                                                                                      
10  Harbourfront East , Union Station , Toronto Islands                                                                
11  Little Portugal , Trinity                                                                                          
12  The Danforth West , Riverdale                                                                                      
13  Toronto Dominion Centre , Design Exchange                                                                          
14  Brockton , Parkdale Village , Exhibition Place                                                                     
15  India Bazaar , The Beaches West                                                                                    
16  Commerce Court , Victoria Hotel                                                                                    
17  Studio District                                                                                                    
20  Davisville North                                                                                                   
21  Forest Hill North & West                                                                                           
22  High Park , The Junction South                                                                                     
23  North Toronto West                                                                                                 
24  The Annex , North Midtown , Yorkville                                                                              
25  Parkdale , Roncesvalles                                                                                            
26  Davisville                                                                                                         
27  University of Toronto , Harbord                                                                                    
28  Runnymede , Swansea                                                                                                
30  Kensington Market , Chinatown , Grange Park                                                                        
31  Summerhill West , Rathnelly , South Hill , Forest Hill SE , Deer Park                                              
32  CN Tower , King and Spadina , Railway Lands , Harbourfront West , Bathurst Quay , South Niagara , Island airport   
34  Stn A PO Boxes                                                                                                     
35  St. James Town , Cabbagetown                                                                                       
36  First Canadian Place , Underground city                                                                            
37  Church and Wellesley

Cluster 2:

In [32]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
29  Moore Park , Summerhill East  1               Restaurant             

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29  Playground            Dance Studio          Electronics Store      

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
29  Eastern European Restaurant  Dumpling Restaurant   Donut Shop             

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29  Doner Restaurant      Dog Run               Distribution Center

Cluster 3: 

In [36]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
33  Rosedale     2               Park                  Trail                  

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
33  Playground            Dance Studio          Electronics Store      

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
33  Eastern European Restaurant  Dumpling Restaurant   Donut Shop             

   9th Most Common Venue 10th Most Common Venue  
33  Doner Restaurant      Dog Run

Cluster 4:

In [34]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19  Roselawn     3               Home Service          Garden                 

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19  Women's Store         Department Store      Ethiopian Restaurant   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
19  Electronics Store     Eastern European Restaurant  Dumpling Restaurant    

   9th Most Common Venue 10th Most Common Venue  
19  Donut Shop            Doner Restaurant

Cluster 5:

In [35]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
18  Lawrence Park  4               Park                  Bus Line               

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18  Swim School           Event Space           Ethiopian Restaurant   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
18  Electronics Store     Eastern European Restaurant  Dumpling Restaurant    

   9th Most Common Venue 10th Most Common Venue  
18  Donut Shop            Doner Restaurant